### PR TITLE
Initial port to monorepo build system.

### PR DESCRIPTION
**Pull request by allisonvacanti**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,22 +1,34 @@
...
-# 3.18 for C++17 + CUDA.
+# 3.15 is the minimum for including the project with add_subdirectory.
+# 3.21 is the minimum for the developer build.
 cmake_minimum_required(VERSION 3.15)
 
 # Determine whether libcudacxx is the top-level project or included into
 # another project via add_subdirectory().
-if ("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_LIST_DIR}")
-  set(LIBCUDACXX_TOPLEVEL_PROJECT ON)
+enable_language(CXX) # For CMAKE_CXX_COMPILER_ID
```
Comment (robertmaynard) [inline CMakeLists.txt:7]: Calling `enable_language` before `project` is undefined behavior and shouldn't be done.

File: lib/cmake/libcudacxx/libcudacxx-config-version.cmake [hunk 2]
```diff
@@ -26,7 +26,9 @@ set(PACKAGE_VERSION_EXACT FALSE)
 set(PACKAGE_VERSION_UNSUITABLE FALSE)
 
 if(PACKAGE_VERSION VERSION_GREATER_EQUAL PACKAGE_FIND_VERSION)
-  set(PACKAGE_VERSION_COMPATIBLE TRUE)
+  if(libcudacxx_VERSION_MAJOR VERSION_EQUAL PACKAGE_FIND_VERSION_MAJOR)
```
Comment (robertmaynard) [inline lib/cmake/libcudacxx/libcudacxx-config-version.cmake:29]: I believe that libcudacxx wants compatibiltiy across major version values.

**Review by robertmaynard — CHANGES_REQUESTED** [review]

File: .upstream-tests/test/CMakeLists.txt [hunk 3]
```diff
@@ -8,6 +8,19 @@ endmacro()
 
 set(LIBCUDACXX_HIGHEST_COMPUTE_ARCH 90)
 
+# Initialize the libcu++ target arches from the CMake property. If both are set,
+# favor the libcu++ variable.
+if (DEFINED CMAKE_CUDA_ARCHITECTURES)
+  if (NOT DEFINED LIBCUDACXX_COMPUTE_ARCHS)
+    list(JOIN CMAKE_CUDA_ARCHITECTURES " " LIBCUDACXX_COMPUTE_ARCHS)
```
Comment (allisonvacanti) [inline .upstream-tests/test/CMakeLists.txt:15]: These should not be related. The libcudacxx is semantically different, and only used for advanced usecases. Leave it alone in this patch.

File: .upstream-tests/test/host_only/CMakeLists.txt [hunk 4]
```diff
@@ -39,8 +39,8 @@ function(libcudacxx_add_standalone_header_test test_file_output_var header_under
 endfunction()
 
 # Don't generate CUDA targets, they fail currently.
-file(GLOB cuda_headers     LIST_DIRECTORIES false RELATIVE ${CMAKE_SOURCE_DIR}/include ${CMAKE_SOURCE_DIR}/include/cuda/*)
-file(GLOB cuda_std_headers LIST_DIRECTORIES false RELATIVE ${CMAKE_SOURCE_DIR}/include ${CMAKE_SOURCE_DIR}/include/cuda/std/*)
+file(GLOB cuda_headers     LIST_DIRECTORIES false RELATIVE ${libcudacxx_SOURCE_DIR}/include ${CMAKE_SOURCE_DIR}/include/cuda/*)
```
Comment (allisonvacanti) [inline .upstream-tests/test/host_only/CMakeLists.txt:42]: Path should be quoted

File: lib/cmake/libcudacxx/libcudacxx-config-version.cmake (hunk 2 above)
Comment (allisonvacanti) [reply lib/cmake/libcudacxx/libcudacxx-config-version.cmake:29]: @wmaxey @jrhemstad Come up with a truth table for how requested version and found version should interact for major version 1, 2, and 3+ and I will encode it.

File: CMakeLists.txt (hunk 1 above)
Comment (wmaxey) [reply CMakeLists.txt:7]: I'm not sure any of the hacking we do to make `lit` generate a config is required. We might be able to just 'be normal' and have our project at the top.

File: lib/cmake/libcudacxx/libcudacxx-config-version.cmake (hunk 2 above)
Comment (allisonvacanti) [reply lib/cmake/libcudacxx/libcudacxx-config-version.cmake:29]: Thanks, that's what I thought, but good to have it explicit. Requested in rows, found in columns, right?

Comment (jrhemstad) [reply lib/cmake/libcudacxx/libcudacxx-config-version.cmake:29]: > Requested in rows, found in columns, right?

Yep!

File: CMakeLists.txt (hunk 1 above)
Comment (allisonvacanti) [reply CMakeLists.txt:7]: It looks like we can just move this into the project call once the `CMAKE_CUDA_HOST_COMPILER` hack is removed 👍

**Pull request closed without merging**